### PR TITLE
Backup/restore of icons

### PR DESCRIPTION
**Pull request by rogiervandongen**
Backup/restore of icons

### CODE REVIEW

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ * @copyright &copy; 2022-onwards G J Barnard in respect to modifications of standard topics format.
+ * @copyright &copy; 2024-onwards RvD in respect to modifications for custom icons.
+ * @copyright 2017 Marina Glancy
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class backup_format_vsf_plugin extends backup_format_plugin {
+
+    public function __construct($plugintype, $pluginname, $optigroup, $step) {
+        parent::__construct($plugintype, $pluginname, $optigroup, $step);
+    }
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php]: Why is this needed?

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+/**
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php:17]: Need a blank line, see https://github.com/Lesterhuis-Training-en-Consultancy/moodle-format_vsf/actions/runs/10054515844/job/27789254050?pr=9#step:9:30.

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+        // We'll simply do ALL installed modules.
+        // It will be too much hassle to see what we may wish to include or not,
+        // because this is "high level behaviour". Customizing icons also will not
+        // see what may or may not be wishful. Custom icons is NOT dependent on
+        // which modules are in use in the course.
+        // Customisation can be done for ALL activity types.
+        $courseiconwrapper = new backup_nested_element('courseicons');
+        $courseicon = new backup_nested_element('courseicon', [], ['name']);
+        $sql = 'SELECT DISTINCT m.name FROM {modules} m';
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php:66]: Do all supported databases support 'DISTINCT'?  Would it not be better to use a DB API call instead of pure SQL as recommended (I believe I've seen this in the Moodle developer documentation)?

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        $params = [];
+        $courseicon->set_source_sql($sql, $params);
+
+        $pluginwrapper->add_child($courseiconwrapper);
+        $courseiconwrapper->add_child($courseicon);
+
+        // This seems to work ere, we actually have the old course context available.
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php]: Seems to work?  You don't sound sure.

File: backup/moodle2/restore_format_vsf_plugin.class.php [hunk 5]
```diff
@@ -66,13 +68,28 @@ public function define_course_plugin_structure() {
         if (($target == backup::TARGET_CURRENT_ADDING || $target == backup::TARGET_EXISTING_ADDING) &&
                 $this->need_restore_numsections()) {
             $maxsection = $DB->get_field_sql(
-                'SELECT max(section) FROM {course_sections} WHERE course = ?',
-                [$this->step->get_task()->get_courseid()]);
-            $this->originalnumsections = (int)$maxsection;
+                    'SELECT max(section) FROM {course_sections} WHERE course = ?',
```
Comment (gjb2048) [inline backup/moodle2/restore_format_vsf_plugin.class.php:78]: Ditto with previous SQL comment.

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        // This is not the same as the activity, which holds and stores the original context.
+        // However, this context is not known on the restore process callback.
+        // See the restore class for the rest of the details.
+
+        // Define/get the virtual plugin element with the condition to fulfill.
+        $plugin = $this->get_plugin_element(null, $this->get_format_condition(), 'vsf');
+
+        // Create one standard named plugin element (the visible container) using the base recommended name.
+        $pluginwrapper = new backup_nested_element($this->get_recommended_name());
+        // Connect to container
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php]: Needs full stop.

File: backup/moodle2/backup_format_vsf_plugin.class.php [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        $plugin->add_child($pluginwrapper);
+
+        // Create an element for the modicon. Since half the information I expect is NOT present in restore,
+        // such as the CRITICAL OLD CONTEXT we need for simple file restores, we'll add that here.
+        // Please note we "hardcode" the source data, which we absolutely require to make this work.
+        // GAWD, I hate backup/restore. Unclear, messy, lacking. Just... plain... wrong.
+        // ELOY, YOUR IMPLEMENTATION SUCKS!
+        $customicon = new backup_nested_element('modicon', [], ['itemid', 'modname', 'oldctxid']);
+        $customicon->set_source_array([['itemid' => 0, 'modname' => $this->task->get_modulename(), 'oldctxid' => $modulecontextid]]);
```
Comment (gjb2048) [inline backup/moodle2/backup_format_vsf_plugin.class.php]: Line too long.

File: backup/moodle2/restore_format_vsf_plugin.class.php [hunk 8]
```diff
@@ -106,20 +131,100 @@ public function after_restore_course() {
...
             // in the original course.
             // Exception is when we restore with merging and the course already had a section with this section number,
             // in this case we don't modify the visibility.
             if ($this->step->get_task()->get_setting_value($key . '_included')) {
-                $sectionnum = (int)$section->title;
+                $sectionnum = (int) $section->title;
                 if ($sectionnum > $numsections && $sectionnum > $this->originalnumsections) {
                     $DB->execute("UPDATE {course_sections} SET visible = 0 WHERE course = ? AND section = ?",
-                        [$this->step->get_task()->get_courseid(), $sectionnum]);
+                            [$this->step->get_task()->get_courseid(), $sectionnum]);
```
Comment (gjb2048) [inline backup/moodle2/restore_format_vsf_plugin.class.php]: Why the extra indentation?

File: backup/moodle2/restore_format_vsf_plugin.class.php [hunk 9]
```diff
@@ -106,20 +131,100 @@ public function after_restore_course() {
...
+        // Add data object to static property for use later (see below).
+        static::$modicons[] = $data;
+    }
+
+    /**
+     * Holds data objects that refer to custom module instance icons.
+     *
+     * @var array
+     */
+    protected static $modicons = [];
```
Comment (gjb2048) [inline backup/moodle2/restore_format_vsf_plugin.class.php]: Should this be defined at the top of the class?  For maintainability and readability.

File: version.php [hunk 10]
```diff
@@ -25,9 +25,9 @@
 
 defined('MOODLE_INTERNAL') || die();
 
-$plugin->version = 2024050801;
+$plugin->version = 2024050802;
```
Comment (gjb2048) [inline version.php]: Why a version bump?  There is nothing that actually needs it here.  Also version bumping in a code change commit makes it more difficult to back port.  Better to version bump a new development sequence, commit, then add code required in separate commits.  Then even if code needs the version to be bumped then when the client installs the tagged release, this happens.  Thus leaving only devs to cope, which they can by down then upgrading.

**Review by gjb2048 — CHANGES_REQUESTED** [review]
@rogiervandongen Please see my comments and make changes as appropriate.  I do appreciate that there is pure SQL already there and in the Moodle code but my understanding is that it should be avoided if at all possible.  In addition there are other separate issues outside of the context of the pull that appear to be causing Code Checker to complain.

**Comment by rogiervandongen** [comment]
@gjb2048 
Changes made.
Moodle's coding standards become more rediculous with every day passing. World uses PSR standards, moodle decides differently. Blank spaces, new lines... yada yada.
Some things are caused by auto formatting btw (such as "double indentations").

If a database wouldn't even be able to use the most basic SQL definitions, they cannot realistically be considered a usable database. This goes for DISTINCT (which is heavily used in Moodle's core), MIN/MAX functions, etc etc. So yeah, they're supported. By every database Moodle supports (which aren't that many).

Version bump removed.

Last but not least, added "global $DB" in vsf_after_restore_course_numsections() as it was missing.

**Comment by gjb2048** [comment]
@rogiervandongen Thanks, but where is the updated commit please?  Fair enough with the SQL reply, I just did have to ask the question.  And totally with the coding standards, drives me nuts some days!

**Comment by rogiervandongen** [comment]
@gjb2048 
My bad! I'm still getting used to the renewed interfacing of gitkraken.
If I'm not mistaken the 2nd commit should now be synced to the PR.
Cheers :)

Frankly, I've never seen that many non-sensical coding standards which the company that defines them is the exact same company that circumvents nearly every one of them :joy: "The joys of Moodle" (one of the worst to me is forbidding protected variables that are "underscore prefixed". The so called persistence model could be so much easier. From the moment I've started coding in PHP, which was somewhere in '07, all my business objects and models depended on "protected $_a", with getters and setters that then rely on magic __get, __set, __isset and __unset implementations that make for a much better to understand implementation. It also makes PHP's new Attributes implementation more than just easy to facilitate, aside from the fact actually implementing any logic using those attributes is terribly slow, unfortunately).
Anyway, those are not really discussions to be held here in a github repository ;)

**Comment by gjb2048** [comment]
@rogiervandongen Thanks, merged, the actual 'release' in version.php does need to go back, but I can do that as it seems I also have some code checker things to do on my code!  Grrr!  And, yeah with coding standards, nuff said!
